### PR TITLE
Setup Stripe webhooks

### PR DESCRIPTION
**Pull request by e-wai**
[1/2] https://www.notion.so/uwblueprintexecs/Task-Board-F22-9e03bf9f7c4343f09a8a5c16dc48a12a?p=7f51d5e490d0494ea2512dd2b463186d&pm=c

- Need to fix the camper charges if there was a coupon used

## Implementation description
* Right now, we send confirmation email in the "create campers" request, and then confirm payment as a separate request. The confirmation email (as well as sending the waiver eventually) should be sent after payment is completed
  * We do have to figure out how to handle incomplete checkouts -- should they continue holding capacity in the camp forever?
* WHEN THIS GETS MERGED TO STAGING: Enable the backend endpoint in the [test Stripe dashboard](https://dashboard.stripe.com/test/webhooks)
* WHEN THIS GETS MERGED TO MAIN: Add the backend endpoint in the production Stripe dashboard to listen to `checkout.session.completed` events (I think FON has credentials, I don't have access)

## Steps to test
1. Go through payments flow, check that you've (1) received the email, (2) it's been marked as `hasPaid` in Mongo. For local testing, download the Stripe CLI, `stripe login` and forward using `stripe listen --forward-to localhost:5000/campers/confirm-payment`

`stripe trigger checkout.session.completed` is another useful test command

## What should reviewers focus on?
* Anyone have load concerns?
* Various docs to know:
  * https://stripe.com/docs/webhooks
  * https://stripe.com/docs/webhooks/quickstart
  * https://stripe.com/docs/cli/trigger#trigger-event
  * https://stripe.com/docs/api/checkout/sessions/object
  * https://stripe.com/docs/api/events/object

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/typescript/rest/camperRoutes.ts [hunk 1]
```diff
@@ -23,6 +23,9 @@ const camperRouter: Router = Router();
 
 const camperService: ICamperService = new CamperService();
 
+// TODO: secure stripe keys
+const STRIPE_ENDPOINT_KEY = process.env.STRIPE_ENDPOINT_SECRET || "";
```
Comment (rchang0501) [inline backend/typescript/rest/camperRoutes.ts:27]: are we going to implement request verification in another PR or is this a TODO item for this one? cuz from my understanding we need to use this to verify that the requests we receive at the webhook endpoint are legit for security

File: backend/typescript/rest/camperRoutes.ts [hunk 2]
```diff
@@ -115,16 +118,25 @@ camperRouter.get("/:chargeId/:sessionId", async (req, res) => {
 });
 
 // ROLES: unprotected
-/* On successful payment, mark camper as paid */
-camperRouter.post("/confirm-payment/:chargeId", async (req, res) => {
-  const { chargeId } = req.params;
+/* Initiated by Stripe webhook. On successful payment, mark camper as paid. */
+camperRouter.post("/confirm-payment", async (req, res) => {
   try {
```
Comment (rchang0501) [inline backend/typescript/rest/camperRoutes.ts:123]: same comment as earlier - shouldn't we implement the webhook verification here for security or am I missing something?

File: backend/typescript/rest/camperRoutes.ts (hunk 1 above)
Comment (e-wai) [reply backend/typescript/rest/camperRoutes.ts:27]: Yeah, we'll need to do both. I just put it as its own PR

**Review by BraydenRoyston — APPROVED** [review]
Looking good. Tests work locally when using the stripe login + local listener. Enabled webhook on the Stripe dashboard. Merging.

**Pull request merged**